// @azure/data-tables/dist-esm/src/generated/models/index.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
/** Known values of {@link OdataMetadataFormat} that the service accepts. */
export var KnownOdataMetadataFormat;
(function (KnownOdataMetadataFormat) {
    KnownOdataMetadataFormat["ApplicationJsonOdataNometadata"] = "application/json;odata=nometadata";
    KnownOdataMetadataFormat["ApplicationJsonOdataMinimalmetadata"] = "application/json;odata=minimalmetadata";
    KnownOdataMetadataFormat["ApplicationJsonOdataFullmetadata"] = "application/json;odata=fullmetadata";
})(KnownOdataMetadataFormat || (KnownOdataMetadataFormat = {}));
/** Known values of {@link ResponseFormat} that the service accepts. */
export var KnownResponseFormat;
(function (KnownResponseFormat) {
    KnownResponseFormat["ReturnNoContent"] = "return-no-content";
    KnownResponseFormat["ReturnContent"] = "return-content";
})(KnownResponseFormat || (KnownResponseFormat = {}));
/** Known values of {@link GeoReplicationStatusType} that the service accepts. */
export var KnownGeoReplicationStatusType;
(function (KnownGeoReplicationStatusType) {
    KnownGeoReplicationStatusType["Live"] = "live";
    KnownGeoReplicationStatusType["Bootstrap"] = "bootstrap";
    KnownGeoReplicationStatusType["Unavailable"] = "unavailable";
})(KnownGeoReplicationStatusType || (KnownGeoReplicationStatusType = {}));
