// @azure/cosmos/dist-esm/src/index.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
export { DEFAULT_PARTITION_KEY_PATH } from "./common/partitionKeys";
export { StatusCodes } from "./common";
export { extractPartitionKey } from "./extractPartitionKey";
export { setAuthorizationTokenHeaderUsingMasterKey } from "./auth";
export { BulkOperationType, } from "./utils/batch";
export { PatchOperationType, } from "./utils/patch";
export { ConnectionMode, ConsistencyLevel, DatabaseAccount, DataType, IndexingMode, SpatialType, GeospatialType, IndexKind, PermissionMode, TriggerOperation, TriggerType, UserDefinedFunctionType, } from "./documents";
export { Constants, OperationType, ResourceType, HTTPMethod } from "./common";
export * from "./request";
export { QueryIterator } from "./queryIterator";
export * from "./queryMetrics";
export { CosmosClient } from "./CosmosClient";
export * from "./client";
export { Scripts } from "./client/Script/Scripts";
export { PluginOn } from "./plugins/Plugin";
export { ChangeFeedIterator } from "./ChangeFeedIterator";
export { ChangeFeedResponse } from "./ChangeFeedResponse";
export { ClientContext } from "./ClientContext";
export { GlobalEndpointManager } from "./globalEndpointManager";
export { SasTokenPermissionKind } from "./common/constants";
export { createAuthorizationSasToken } from "./utils/SasToken";
export { RestError } from "@azure/core-rest-pipeline";
export { AbortError } from "@azure/abort-controller";
